reco_utils/common/tf_utils.py

```diff
@@ -0,0 +1,408 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+import abc
+import itertools
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+
+from reco_utils.common.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL
+)
+
+
+MODEL_DIR = 'model_checkpoints'
+
+
+def pandas_input_fn(
+    df,
+    y_col=None,
+    batch_size=128,
+    num_epochs=1,
+    shuffle=False,
+    num_threads=1
+):
+    """Pandas input function for TensorFlow high-level API Estimator.
+
+    tf.estimator.inputs.pandas_input_fn cannot handle array/list column properly.
+    If the df does not include any array/list data column, one can simply use TensorFlow's pandas_input_fn.
+
+    For more information, see (https://www.tensorflow.org/api_docs/python/tf/estimator/inputs/numpy_input_fn)
+
+    Args:
+        df (pd.DataFrame): Data containing features
+        y_col (str): Label column name if df has it.
+        batch_size (int): Batch size for the input function
+        num_epochs (int): Number of epochs to iterate over data. If None will run forever.
+        shuffle (bool): If True, shuffles the data queue.
+        num_threads (int): Number of threads used for reading and enqueueing.
+
+    Returns:
+        (tf.estimator.inputs.numpy_input_fn) function that has signature of ()->(dict of features, targets)
+    """
+
+    X_df = df.copy()
+    if y_col is not None:
+        y = X_df.pop(y_col).values
+    else:
+        y = None
+
+    X = {}
+    for col in X_df.columns:
+        values = X_df[col].values
+        if isinstance(values[0], (list, np.ndarray)):
+            values = np.array([l for l in values], dtype=np.float32)
+        X[col] = values
+
+    input_fn = tf.estimator.inputs.numpy_input_fn(
+        x=X,
+        y=y,
+        batch_size=batch_size,
+        num_epochs=num_epochs,
+        shuffle=shuffle,
+        num_threads=num_threads
+    )
+
+    return input_fn
+
+
+def _build_optimizer(name, lr=0.001, ftrl_l1_reg=0.0):
+    """Get an optimizer for TensorFlow high-level API Estimator.
+
+    Args:
+        name (str): Optimizer name.
+        lr (float): Learning rate.
+        ftrl_l1_reg (float): Ftrl optimizer l1 regularization rate. Ignored for other optimizers.
+
+    Returns:
+        optimizer (tf.train.Optimizer)
+    """
+    if name == 'Adagrad':
```

Comment:
what about sgd with momentum?

Reply:
Adding couple more optimizers in TF including Momentum to the next commit.